File: scripts/build_blog_index.py
from __future__ import annotations

import json
import re
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BLOGS_DIR = ROOT / "blogs"
OUTPUT_PATH = BLOGS_DIR / "index.json"
HEADING_PATTERN = re.compile(r"^\s*#\s+(.+)$")


def derive_title(markdown_path: Path) -> str:
    try:
        with markdown_path.open("r", encoding="utf-8") as handle:
            for line in handle:
                line = line.lstrip("\ufeff")
                match = HEADING_PATTERN.match(line)
                if match:
                    return match.group(1).strip()
    except OSError:
        pass
    slug = markdown_path.stem.replace('-', ' ').replace('_', ' ')
    words = [word.capitalize() for word in slug.split() if word]
    return " ".join(words) if words else markdown_path.stem


def build_manifest() -> dict:
    posts = []
    for path in sorted(BLOGS_DIR.glob("*.md")):
        if not path.is_file():
            continue
        stat = path.stat()
        title = derive_title(path)
        posts.append({
            "path": f"blogs/{path.name}",
            "filename": path.name,
            "title": title,
            "lastModified": datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc).isoformat(),
            "size": stat.st_size,
            "_sort": stat.st_mtime,
        })
    posts.sort(key=lambda item: (item["_sort"], item["filename"]), reverse=True)
    for post in posts:
        post.pop("_sort", None)
    return {
        "generatedAt": datetime.now(tz=timezone.utc).isoformat(),
        "posts": posts,
    }


def main() -> None:
    BLOGS_DIR.mkdir(parents=True, exist_ok=True)
    manifest = build_manifest()
    payload = json.dumps(manifest, indent=2) + "\n"
    if OUTPUT_PATH.exists():
        OUTPUT_PATH.unlink()
    OUTPUT_PATH.write_text(payload, encoding="utf-8")
    print(f"Wrote {len(manifest['posts'])} posts to {OUTPUT_PATH.relative_to(ROOT)}")


if __name__ == "__main__":
    main()

File: scripts/build-blog-index.mjs
import { promises as fs } from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const rootDir = path.resolve(__dirname, "..");
const blogsDir = path.join(rootDir, "blogs");
const outputPath = path.join(blogsDir, "index.json");

function slugToTitle(slug) {
  return slug
    .split(/[-_]+/)
    .filter(Boolean)
    .map(part => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ") || slug;
}

async function extractHeading(filePath) {
  try {
    const raw = await fs.readFile(filePath, "utf8");
    const lines = raw.split(/\r?\n/);
    for (const line of lines) {
      const cleaned = line.replace(/^\uFEFF/, "");
      const match = /^\s*#\s+(.+)$/.exec(cleaned);
      if (match) {
        return match[1].trim();
      }
    }
  } catch (error) {
    console.warn(`Could not read ${filePath}: ${error.message}`);
  }
  return null;
}

async function buildIndex() {
  const entries = [];
  const dirItems = await fs.readdir(blogsDir, { withFileTypes: true });

  for (const item of dirItems) {
    if (!item.isFile() || !item.name.toLowerCase().endsWith(".md")) {
      continue;
    }
    const filePath = path.join(blogsDir, item.name);
    const stat = await fs.stat(filePath);
    const heading = await extractHeading(filePath);
    const filename = item.name;
    const slug = filename.replace(/\.md$/i, "");

    entries.push({
      path: `blogs/${filename}`,
      filename,
      title: heading || slugToTitle(slug),
      lastModified: stat.mtime.toISOString(),
      size: stat.size,
      sortKey: stat.mtimeMs
    });
  }

  entries.sort((a, b) => {
    if (b.sortKey !== a.sortKey) {
      return b.sortKey - a.sortKey;
    }
    return a.filename.localeCompare(b.filename);
  });

  const manifest = {
    generatedAt: new Date().toISOString(),
    posts: entries.map(({ sortKey, ...rest }) => rest)
  };

  await fs.writeFile(outputPath, `${JSON.stringify(manifest, null, 2)}\n`, "utf8");
  console.log(`Wrote ${entries.length} posts to ${path.relative(rootDir, outputPath)}`);
}

buildIndex().catch(error => {
  console.error(error);
  process.exitCode = 1;
});
